src/integration/java/org/springframework/data/couchbase/core/CouchbaseTemplateTests.java

```diff
@@ -402,6 +422,67 @@ public void shouldLoadVersionPropertyOnFind() throws Exception {
 		assertEquals(versionedClass.getVersion(), foundClass.getVersion());
 	}
 
+	@Test
+	public void shouldUpdateAlreadyExistingDocument() throws Exception {
+		final String key = testName.getMethodName();
+		removeIfExist(key);
+
+		final AtomicLong counter = new AtomicLong();
+
+		VersionedClass initial = new VersionedClass(key, "value-0");
+		template.save(initial);
+
+		AsyncUtils.executeConcurrently(3, new Callable<Void>() {
+			@Override
+			public Void call() throws Exception {
+				boolean saved = false;
+				while(!saved) {
+					long counterValue = counter.incrementAndGet();
+					VersionedClass messageData = template.findById(key, VersionedClass.class);
+					messageData.field = "value-" + counterValue;
+					try {
+						template.save(messageData);
+						saved = true;
+					} catch (OptimisticLockingFailureException e) {
+					}
```

Comment:
all saves will succeed here since all of them are retried with updated version. Maybe you are talking about another test case?